Mathlib/NumberTheory/Cyclotomic/Three.lean

```diff
@@ -0,0 +1,115 @@
+`λ = η - 1`.
+
+## Main results
+* `IsCyclotomicExtension.Rat.Three.Units.mem`: Given a unit `u : (𝓞 K)ˣ`, we have that
+`u ∈ {1, -1, η, -η, η^2, -η^2}`.
+
+* `IsCyclotomicExtension.Rat.Three.eq_one_or_neg_one_of_unit_of_congruent`: Given a unit
+`u : (𝓞 K)ˣ`, if `u` is congruent to an integer modulo `3`, then `u = 1` or `u = -1`.
+
+This is a special case of the so-called *Kummer's lemma* (see for example
+https://kconrad.math.uconn.edu/blurbs/gradnumthy/kummer.pdf).
+-/
+
+open NumberField Units InfinitePlace nonZeroDivisors Polynomial
+
+namespace IsCyclotomicExtension.Rat.Three
+
+variable {K : Type*} [Field K] [NumberField K] [IsCyclotomicExtension {3} ℚ K]
+variable {ζ : K} (hζ : IsPrimitiveRoot ζ ↑(3 : ℕ+)) (u : (𝓞 K)ˣ)
+local notation3 "η" => hζ.toInteger
+local notation3 "λ" => hζ.toInteger - 1
+
+/-- Let `u` be a unit in `(𝓞 K)ˣ`, then `u ∈ {1, -1, η, -η, η^2, -η^2}`. -/
+theorem Units.mem : ↑u ∈ ({1, -1, η, -η, η ^ 2, -η ^ 2} : Set (𝓞 K)) := by
+  have hrank : rank K = 0 := by
+    dsimp only [rank]
+    rw [card_eq_nrRealPlaces_add_nrComplexPlaces, nrRealPlaces_eq_zero (n := 3) K (by decide),
+      zero_add, nrComplexPlaces_eq_totient_div_two (n := 3)]
+    rfl
+  obtain ⟨⟨x, e⟩, hxu, -⟩ := exist_unique_eq_mul_prod _ u
+  replace hxu : u = x := by
+    rw [← mul_one x.1, hxu]
+    apply congr_arg
+    rw [← Finset.prod_empty]
+    congr
+    rw [Finset.univ_eq_empty_iff, hrank]
+    infer_instance
+  obtain ⟨n, hnpos, hn⟩ := isOfFinOrder_iff_pow_eq_one.1 <| (CommGroup.mem_torsion _ _).1 x.2
+  replace hn : (↑u : K) ^ ((⟨n, hnpos⟩ : ℕ+) : ℕ) = 1 := by
+    rw [← map_pow]
+    convert map_one (algebraMap (𝓞 K) K)
+    rw_mod_cast [hxu, hn]
+    simp
+  obtain ⟨r, hr3, hru⟩ := hζ.exists_pow_or_neg_mul_pow_of_isOfFinOrder (by decide)
+    (isOfFinOrder_iff_pow_eq_one.2 ⟨n, hnpos, hn⟩)
+  replace hr : r ∈ Finset.Ico 0 3 := Finset.mem_Ico.2 ⟨by simp, hr3⟩
+  replace hru : ↑u = η ^ r ∨ ↑u = -η ^ r := by
+    rcases hru with (h | h)
+    · left; ext; exact h
+    · right; ext; exact h
+  fin_cases hr
+  all_goals{
+    rcases hru with (h | h)
+    · simp only [h, pow_zero, Set.mem_insert_iff, one_ne_zero, Set.mem_singleton_iff, false_or,
+        true_or, zero_add, pow_one, eq_neg_self_iff, true_or, or_true]
+    · simp [h, pow_zero, Set.mem_insert_iff, neg_eq_self_iff, one_ne_zero, Set.mem_singleton_iff,
+        true_or, or_true, zero_add, pow_one, neg_inj, neg_eq_self_iff]}
+
+/-- We have that `λ ^ 2 = -3 * η`. -/
+lemma lambda_sq : λ ^ 2 = -3 * η :=
+  calc λ ^ 2 = η ^ 2 + η + 1 - 3 * η := by ring
+  _ = 0 - 3 * η := by ext; simpa using hζ.isRoot_cyclotomic (by decide)
+  _ = -3 * η := by ring
+
+/-- If a unit `u` is congruent to an integer modulo `λ ^ 2`, then `u = 1` or `u = -1`.
+
+This is a special case of the so-called *Kummer's lemma*. -/
+theorem eq_one_or_neg_one_of_unit_of_congruent (hcong : ∃ n : ℤ, λ ^ 2 ∣ (u - n : 𝓞 K)) :
+    u = 1 ∨ u = -1 := by
```

Comment:
The proof looks unnecessarily involved to me. If you know that `(𝓞 K)ˣ = {1,-1,ζ,-ζ,ζ+1,-ζ-1}`, then this follows from `3 ∣ a + b*ζ → 3 ∣b`, without the need to use the general result `not_exists_int_prime_dvd_sub_of_prime_ne_two'` (which, of course, will still be useful for the general case).

Reply:
I am not sure I understand this comment, why `3 ∣ a + b*η → 3 ∣b`?

Reply:
I mean when `a` and `b` are (rational) integers.

Reply:
Since `[1, η]` is a `ℤ`-basis of `𝓞 K`, an element `a + b*η` (with `a b : ℤ`) is divisible by an integer `n` if and only if `n` divides both `a` and `b`.

Reply:
But this would need another lemma and is not really necessary here, as the approach using the general statement works (and the proof is a bit shorter now).